src/access/Subjects.h

```diff
@@ -0,0 +1,208 @@
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include <limits>
+
+#include <access/AuthMode.h>
+#include <lib/core/DataModelTypes.h>
+#include <lib/core/InPlace.h>
+#include <lib/core/NodeId.h>
+#include <lib/support/Variant.h>
+
+namespace chip {
+namespace Access {
+
+struct PaseSubject
+{
+    static constexpr uint16_t kPasscodeId = 0;
+    bool operator==(const PaseSubject & that) const { return true; }
+};
+
+struct NodeSubject
+{
+    NodeSubject(NodeId aNodeId) : nodeId(aNodeId) {}
+    NodeId nodeId;
+    bool operator==(const NodeSubject & that) const { return this->nodeId == that.nodeId; }
+};
+
+struct GroupSubject
+{
+    GroupSubject(GroupId aGroupId) : groupId(aGroupId) {}
+    GroupId groupId;
+    bool operator==(const GroupSubject & that) const { return this->groupId == that.groupId; }
+};
+
+class OperationalNodeId;
+
+/** A scoped subject is a unique identifier with-in a fabric scope. */
+class ScopedSubject
+{
+public:
+    ScopedSubject() {}
+
+    template <typename T, class... Args>
+    constexpr explicit ScopedSubject(InPlaceTemplateType<T>, Args &&... args) :
+        mSubject(InPlaceTemplate<T>, std::forward<Args>(args)...)
+    {}
+
+    template <typename T, typename... Args>
+    static ScopedSubject Create(Args &&... args)
+    {
+        return ScopedSubject(InPlaceTemplate<T>, std::forward<Args>(args)...);
+    }
+
+    Access::AuthMode GetAuthMode() const
+    {
+        if (mSubject.Is<PaseSubject>())
+        {
+            return Access::AuthMode::kPase;
+        }
+        else if (mSubject.Is<NodeSubject>())
+        {
+            return Access::AuthMode::kCase;
+        }
+        else if (mSubject.Is<GroupSubject>())
+        {
+            return Access::AuthMode::kGroup;
+        }
+        else
+        {
+            return Access::AuthMode::kNone;
+        }
+    }
+
+#if CHIP_DETAIL_LOGGING
+    const char * GetAuthModeString() const
+    {
+        switch (GetAuthMode())
+        {
+        case Access::AuthMode::kPase:
+            return "Pase";
+        case Access::AuthMode::kCase:
+            return "Case";
+        case Access::AuthMode::kGroup:
+            return "Group";
+        case Access::AuthMode::kNone:
+        default:
+            return "None";
+        }
+    }
+#endif // CHIP_DETAIL_LOGGING
+
+    /** Return subject value described in spec */
+    uint64_t GetValue() const
```

Comment:
This is unused outside logging.  And I don't see how one would use it, exactly, without the authmode...

If this is just meant for logging, it should be named accordingly and probably `#if CHIP_DETAIL_LOGGING`.  If it's meant to be used in some other way, it needs very clear documentation about how to use it, and probably still a better name.

Reply:
It can be used when encoding a subject into commands, like `AddNOC` command.

I'll change the name to `GetSubjectId` to make it aligned with spec.

Reply:
AddNOC does not really take a subject.  It claims to, but in practice the subject there is always a node id.  Has to be, given how it gets used.  I wish it just clearly said it's a node-id instead of using SubjectID....  @mlepage-google @tcarmelveilleux 

But more to the point: when would one ever use a Subject as defined here to encode an AddNOC command?

Reply:
In spec, it says `CaseAdminNode` is a `SubjectId`

For Groups and CATs, I have provided `GroupSubject` and `CaseAuthenticatedTagSubject` to help constructing the uint64 value.

It is better to use the given subject classes rather than manually manipulate the CAT bits and construct `SubjectId`.

Reply:
> In spec, it says CaseAdminNode is a SubjectId

OK.  So which SubjectIds would actually be allowed there?

> It is better to use the given subject classes rather than manually manipulate the CAT bits and construct SubjectId

Yes, but why would someone "manually manipulate the CAT bits" instead of using the existing utility we have for this?